### DISCUSSION TITLE
Config precendce unclear

### DISCUSSION THREAD

**Discussion by KUGA2** [renovatebot/renovate#32598] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Azure

### Please tell us more about your question or problem

Hi.

I am struggling with config precedence. I think nearly all Info is written down in [Renovate configuration overview](https://docs.renovatebot.com/config-overview/#renovate-configuration-overview) but it lacks an overview (similar to [here ](https://github.com/renovatebot/renovate/issues/7430#issuecomment-1006506931) but not done I think).

The Order I think it is loaded and overwritten:

- Default config (hard coded defaults)
- Global settings
  - config,js
  - environment variables
  - cli
- Repo settings (renovate.json etc)

I think it should be clearly documented as a list like that, not in "Headline Order" or in the section text.

PS: You wanted initially wanted to commit that but original issue went in a different direction.

Special unclearity: What about list configs like baseBranches? Will they be completely overwritten or list elements appended?

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by rarkins** [comment]
> I think it should be clearly documented as a list like that, not in "Headline Order" or in the section text. 

The headlines are also listed at the top:

![image](https://github.com/user-attachments/assets/03e92b81-427c-4af3-8edf-00d024f35325)

> That said, I think it is not true for "token".

The same is true for any global config. By definition it is global-only and not available in repository config.

> Special unclearity: What about list configs like baseBranches? Will they be completely overwritten or list elements appended?

Please describe your question in more words. Overwritten by what, for example?

**Comment by KUGA2** [comment]
Thanks,

1. The list is there, but it does not say thats also the order of precedence! Also it does not state the order of the 3 possible global options (env, cli, config).
2. Dont mind the token. It seems I forget to delete the sentence. Token can only be set in global config not in repo config. Thats why overwriting from repo config did not work. This is documented at token. (Removed it from original post)
3. If global hat baseBranches={b1, b2} and repo hat baseBranches={b3, b4} what will be the result? Only the latter or all 4 branches? Is that the same for all list config elements?

**Reply by rarkins** [reply]
1. You're welcome to submit a PR improving the docs if you wish
2. OK
3. `baseBranches` is not mergeable, which means the result will be b3,b4. Config elements which are mergeable will indicate it on https://docs.renovatebot.com/configuration-options/

**Reply by KUGA2** [reply]
What about:

```
When Renovate runs on a repository, the final config used is derived from the:

- Default config
- Global config
  - File config
  - Environment config
  - CLI config
- Inherited config
- Repository config
- Resolved presets referenced in config

This order is the order of precedence. Later configuration overwrites or merges with previous ones, depending whether the element is marked as mergable.

<!-- prettier-ignore -->
!!! note
    If you use a Mend-hosted app, please read [Mend-hosted Apps Configuration](mend-hosted/hosted-apps-config.md) _after_ reading this page.
```

or do you prefer it in a note/tipps section? (I am not sure how that would work syntaxwise)

**Reply by viceice** [reply]
repo config overrides resolved presets

**Reply by KUGA2** [reply]
> repo config overrides resolved presets

Are you saying the current doc is wrong?

**Reply by rarkins** [reply]
The order of the last two points is wrong

**Reply by KUGA2** [reply]
If its wrong in my post its also wrong in current master. I will create a PR.
